### DISCUSSION TITLE
commit callback or message callback and --preserve-commit-hashes not working?

### DISCUSSION THREAD

**Discussion by devkoa** [newren/git-filter-repo#198] [General]
I'm trying to update commit messages and keep the same hash.

i try both options --message-callback and --commit-callback but no mater which one i choose, it generates new hashes. Here how i do that:

python3 git-filter-repo.py --preserve-commit-hashes --message-callback (or --commit-callback) '
if b"blabla" not in message:
   message = b"MyMessage " + message
return message' --force

Is this a kind of bug? Or i do something completely wrong?

Any help is appreciated

**Comment by newren** [comment]
--preserve-commit-hashes is only about leaving commit hashes *within commit messages* alone.  You cannot leave the commit hashes themselves alone by git's design.  The commit hashes are a hash of the several things, including the commit message.  So, this isn't a bug, and you aren't doing anything wrong, you're just asking for something that is impossible within git's design.

If you really need to preserve commit hashes, then rewriting history is not an option for you.  You could look into git notes if you want to provide an updated annotation of sorts that provides additional information to correct mistakes in commit messages.

**Reply by devkoa** [reply]
thanks for answering
